### PR TITLE
doc: added contributing.md

### PR DESCRIPTION
**Pull request by thelostone-mc**
Set up a base template to provide some clarity to encourage fresh
faces to contribute.
Refs: https://github.com/gitcoinco/web/issues/65

Modules(s) changed:
-  doc

### CODE REVIEW

**Comment by thelostone-mc** [comment]
@mbeacom @owocki So I made a base draft. Have a look into it and i'll work on the feedback.

You might want to play close attention to the [commit](https://github.com/gitcoinco/web/pull/97/files#diff-6a3371457528722a734f3c51d9238c13R78) & [code](https://github.com/gitcoinco/web/pull/97/files#diff-6a3371457528722a734f3c51d9238c13R69) sections

I've made a few assumptions, we can talk it over if not everyone is okay with it ^_^
A good amount of this is inspired from node contributing.md and I loved it when I had come across it the very first time

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+Contributors can disagree with one another so long as they are done in good
+faith and everyone is working towards a common goal.
+
+## Issues
+Issues in `gitcoin/web` are the primary means by which bug reports and
+general discussions are made. An contributor is allowed to create an issue,
+discuss and provide a fix if needed.
+
+## Discussions And General Help
+As Gitcoin is still at it's early stages, drop by gitcoincommunity.slack.com
```
Comment (owocki) [inline CONTRIBUTING.md]: gitcoin.co/slack is where one can request an invite.. probably better to send folks here.

**Comment by owocki** [comment]
Should we add anything about liscensing?  Is that standard?

**Comment by owocki** [comment]
This seems like a good addition to me.  Doesnt seem too daunting.  If we merge this in we should update the other gitcoin repos too.  Whats everyone else think?

**Comment by thelostone-mc** [comment]
@owocki We could link to the `license.md` which already exists in the repo, we could add a link at the bottom but is it needed ? I'm in two minds

Are you sure you & @mbeacom  had no problem with the commit format ? 😛  I assumed there would be a few counters here !
I enforced it at office and folks hate me for it when I reject PRs cause of the format ( but it makes things looks nicer and they eventually had to gave in )

**Comment by mbeacom** [comment]
I regularly use a linter and push pep8 for python dev.  `2` (html) and `4` (py) makes sense to me. pep8 and all.  I do disagree with `Wrap all other lines at 72 columns.` - With modern screens and all, wrapping at 72 is dying off. `120` is a common best practice now, but I can live with it.

I think the commit message outline is way too strict for something hoping to engage the open source community to actively commit changes. It makes sense for internal project with teams working together regularly, but for OS? I think a process and outline are great, but that might be taking it too far.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+   __Examples:__
+   - `Fixes: https://github.com/gitcoinco/web/issues/87`
+   - `Refs: https://github.com/gitcoinco/web/issues/91`
+
+5. Sample commit
+   ```txt
+   subsystem: explain the commit in one line
+
+   - list out your changes as points if there are many changes
```
Comment (mbeacom) [inline CONTRIBUTING.md]: If there are many changes, they should probably be split off into separate PRs.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+1. The first line should:
+   - contain a short description of the change (preferably 50 characters or less
+     and no more than 72 characters)
+   - be prefixed with the name of the changed subsystem and start with an
+   imperative verb.
+   - If your commit can be explained in one line, then you can skip whats below
+
+   __Examples:__
+   - `doc: updated README.md`
+   - `ui: added design for testimonial page`
```
Comment (mbeacom) [inline CONTRIBUTING.md]: You're going to have people attempting to submit PRs to fix things that haven't read this before pushing to their branch.  The commit message is irrelevant to me, as long as the PR itself is well documented...  If anything, we should add a PR template.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+easier for folks to understand the gist of it without before jumping to the
+the code / doc.
+
+As a best practice, once you have committed your changes, it is a good idea
+to use `git rebase` (not `git merge`) to ensure your changes are placed at the
+top. Plus merge conflicts can be resolved
+
+```text
+$ git fetch upstream
+$ git rebase upstream/master
```
Comment (mbeacom) [inline CONTRIBUTING.md]: This will 🔥 if they use reverse merging.

File: CONTRIBUTING.md (hunk 2 above)
Comment (thelostone-mc) [reply CONTRIBUTING.md]: Ah but what if they all correspond to the same issue ? doesn't that make reviewing messier ?
If they can be separated into 2 PRs , by all means we would allow it ^_^ 
I can update it there saying or shoot out multiple PRs if that seems feasible. Would that make things clearer ?

**Comment by thelostone-mc** [comment]
@mbeacom 

Things I agree with you a 💯 percent: 
- need for a linter, that would solve a huge chunk of the problem. Let's use `pep8`
   Might need a `CSS` linter too! I'll work on that and shoot out another PR
- will define a template for the PR,  makes everyones life easier, (probably with this PR )

things I'm not entirely on board with : 
-  relaxing the commit message format cause open source it gives you uniformity.
   Initially when I was asked to follow this in the `node` community, I found it troublesome
   but looking back I'm happy I did that. 

   if it helps i could set up a git hook to validate the commit message as guided [here](http://blog.revunit.com/good-commit-messages-enforcing-git-hooks/)

PS : This [blog by Tim Pope](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) made a lot of sense and I keep telling folks to follow this approach (quite a few get annoyed by that )

That said you are right the commit message are strict, if the majority decides it's an overkill, we can take it off ^_^

reverse merging might things messy 😓 . But rebasing ensures purpose of ensuring that your changes are at the top thus avoiding re-writing timelines.

**Comment by owocki** [comment]
> I think the commit message outline is way too strict for something hoping to engage the open source community to actively commit changes. 

on 2nd thought, i agree with this.  @thelostone-mc i worry that if we are too strict it's going to turn people away.  please feel free to disagree with me on this, but i think that the downside here outweighs the upside

> If anything, we should add a PR template.

this is a really good idea.

**Comment by thelostone-mc** [comment]
Alright how about this, we relax the commit message to list out what all changes are present as points or a paragraph ! ( Each line limiting to 100 words )
Once this picks up a lot of attention And we've got a huge community , then we enforce the stricter rules! Does that sound do-able ?

**Comment by owocki** [comment]
sounds good to me.

anyone else have comments before this gets merged?

**Comment by owocki** [comment]
> If anything, we should add a PR template.

this might be worth doing before we merge.  per https://github.com/blog/2111-issue-and-pull-request-templates

**Comment by thelostone-mc** [comment]
@owocki The templates are already added ^_^ 
I guess that slipped passed you 😝

**Comment by thelostone-mc** [comment]
@mbeacom relaxed the commit guidelines as you had requested and added templates

**Comment by ethikz** [comment]
For committing and at least what about a template for messages or this too strict?  Really it's only for changelog purposes

https://github.com/ethikz/guidelines/blob/master/branching-strategy.md

**Comment by thelostone-mc** [comment]
@ethikz Oh this is nice 😄 but @owocki and @mbeacom made a good point that if we make this too strict , we might end up turning new folks away. I'm up for it but I guess @owocki and @mbeacom would be open to stricter commits once the community grows

**Comment by ethikz** [comment]
@thelostone-mc Right that's what I'm asking if that is too strict?  I'm guessing so based on previous comments but just wanted to verify is all.

**Comment by mbeacom** [comment]
Don't get me wrong, I'm all for easily parsed changelogs and proper commit documentation...  It's a matter of making the process easily approachable this early on.  If we document the process and outline what **will** result in a PR being turned away, I'd be 👌 with it. Following that, it would be a matter of sitting back and waiting to see how the community responds.  If it results in numerous abandoned PRs that could potentially add value to the repository, but we are turning them away for syntactical issues in commits, we know we have a problem with either: Upfront explanation/clear direction or too strict guidelines/standards.

In any case, CHANGELOG for the win.

**Comment by thelostone-mc** [comment]
@owocki consider for merge ?
I think a few folks looked over it and didn't have any complains and we've relaxed the guidelines to a good extent , added template for raising a PR and an Issue

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+   __Sample commit B__
+   ```txt
+   - list out your changes as points if there are many changes
+   - if needed you can also send it across as
+   - all wrapped within 100 lines
+
+   Fixes: https://github.com/gitcoinco/web/issues/87
+   Refs: https://github.com/gitcoinco/web/issues/91
+   ```
+5. Ensure your squash your commit to make our history neater and reviewing
```
Comment (owocki) [inline CONTRIBUTING.md]: do we want to do this? ive worked on teams that have gone etiher way and i dont feel like we've derived a lot of value from this approach.  my 2c

**Comment by owocki** [comment]
doc looks good.  @mbeacom @ethikz @jhaas let me know if any more comments.  if not, i have a reminder set to look in a couple hours and merge

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<!--
```
Comment (Elaniobro) [inline .github/PULL_REQUEST_TEMPLATE.md]: Again, this is awesome first step! Here is what we use at work which has helped immensely. Feel free to take/use ideas. Just sharing possibilities :)

**Description:** 

* 

**Issue(s):** 

* #ISSUE#

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+Thank you for your pull request. Please provide a description above and review
+the requirements below.
+
+Contributors guide: https://github.com/gitcoinco/web/blob/contributing/CONTRIBUTING.md
+-->
+
+##### Checklist
```
Comment (Elaniobro) [inline .github/PULL_REQUEST_TEMPLATE.md]: 👍

File: CONTRIBUTING.md (hunk 5 above)
Comment (thelostone-mc) [reply CONTRIBUTING.md]: Well we aren't forcing them to squash it, but making it neater is always welcome.
I have folks who submit 6 PRs which could be actually just 2 commits!  
It's just about keeping the history neater I guess.

How about rephrasing it to  ?
```
Squashing commit to make the history neater are always welcomed
``` 
@Elaniobro your thoughts?

PS: yeah i noticed the spelling & grammar error. will fix that

File: .github/ISSUE_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
+<!--
```
[the remark this answers is not in the document]
Comment (thelostone-mc) [reply .github/ISSUE_TEMPLATE.md]: I'm stealing parts of this 😝

**Comment by thelostone-mc** [comment]
@owocki Added another commit taking into account the feedback ^_^
Have a look at it whenever you get the time

File: CONTRIBUTING.md (hunk 5 above)
Comment (Elaniobro) [reply CONTRIBUTING.md]: Squash and Merge FTW! Yea, all for it.

Good read: https://chris.beams.io/posts/git-commit/#seven-rules

File: .github/ISSUE_TEMPLATE.md [hunk 9]
```diff
@@ -2,21 +2,24 @@
 Thank you for reporting an issue.
 
 This issue tracker is for bugs and issues found within gitcoin.
-If you require more general support reach out to us at
+If it's general support you need, reach out to us at
```
Comment (Elaniobro) [inline .github/ISSUE_TEMPLATE.md]: stay away from contractions. Also should this read as `its` not `it's` nor 'it is'? My grammar is poor. If the latter than do away with the contraction.

File: .github/ISSUE_TEMPLATE.md [hunk 10]
```diff
@@ -2,21 +2,24 @@
...
 gitcoin.co/slack
 
 Please fill in as much of the template below as you're able.
+Provide a general summary of the issue in the Title above
```
Comment (Elaniobro) [inline .github/ISSUE_TEMPLATE.md]: `Title` should be lowercase IMO

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -107,8 +107,7 @@ So we ask you to ensure all these are met before you shoot out a PR.
    Fixes: https://github.com/gitcoinco/web/issues/87
    Refs: https://github.com/gitcoinco/web/issues/91
    ```
-5. Ensure your squash your commit to make our history neater and reviewing
-   easier.
+5. Squashing your commits to make our history neater is always welcomed.
```
Comment (Elaniobro) [inline CONTRIBUTING.md]: Should probably preface this with instructions on how to access the `Squash & Merge` feature. n00b may not know how to? Also should probably read `Squashing and Merging your...` instead of just `Squashing your...`

File: .github/ISSUE_TEMPLATE.md (hunk 9 above)
Comment (thelostone-mc) [reply .github/ISSUE_TEMPLATE.md]: My bad :P

**Comment by thelostone-mc** [comment]
@Elaniobro Ah I've been dealing with grammar nazis all my life ! 😆 
Changes done + rebased + updated PR

Shout out to @Elaniobro @ethikz @mbeacom  for taking the effort to review ^_^

@owocki looks like we are good to go 👍
